### PR TITLE
OpenMPTarget with CrayClang compiler

### PR DESCRIPTION
**Pull request by rgayatri23**
The PR does the following:
* Fix compile issues that arise due to the use of CrayClang compiler with the OpenMPTarget backend
* Block tests that lead to ICE with the same.

### CODE REVIEW

**Review by crtrott — APPROVED** [review]

**Review by janciesko — APPROVED** [review]

File: algorithms/unit_tests/TestBinSortA.hpp [hunk 1]
```diff
@@ -219,6 +219,9 @@ void test_sort_integer_overflow() {
 }  // namespace BinSortSetA
 
 TEST(TEST_CATEGORY, BinSortGenericTests) {
+#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
```
Comment (masterleinad) [inline algorithms/unit_tests/TestBinSortA.hpp:222]:
```suggestion
// FIXME KOKKOS_COMPILER_CRAY_LLVM
#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
```

File: algorithms/unit_tests/TestBinSortB.hpp [hunk 2]
```diff
@@ -185,6 +185,9 @@ void run_for_rank2() {
 }  // namespace BinSortSetB
 
 TEST(TEST_CATEGORY, BinSortUnsignedKeyLayoutStrideValues) {
+#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
+  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
+#endif
```
Comment (masterleinad) [inline algorithms/unit_tests/TestBinSortB.hpp:190]:
```suggestion
// FIXME KOKKOS_COMPILER_CRAY_LLVM
#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
#endif
```

File: algorithms/unit_tests/TestNestedSort.hpp [hunk 3]
```diff
@@ -386,6 +386,10 @@ void test_nested_sort_by_key(unsigned int N, KeyType minKey, KeyType maxKey,
 }  // namespace NestedSortImpl
 
 TEST(TEST_CATEGORY, NestedSort) {
+#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
+  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
+#endif
```
Comment (masterleinad) [inline algorithms/unit_tests/TestNestedSort.hpp:391]:
```suggestion
// FIXME KOKKOS_COMPILER_CRAY_LLVM
#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
#endif
```

File: algorithms/unit_tests/TestNestedSort.hpp [hunk 4]
```diff
@@ -394,6 +398,10 @@ TEST(TEST_CATEGORY, NestedSort) {
 }
 
 TEST(TEST_CATEGORY, NestedSortByKey) {
+#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
+  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
+#endif
```
Comment (masterleinad) [inline algorithms/unit_tests/TestNestedSort.hpp:403]:
```suggestion
// FIXME KOKKOS_COMPILER_CRAY_LLVM
#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
#endif
```

File: algorithms/unit_tests/TestRandom.hpp [hunk 5]
```diff
@@ -542,6 +542,10 @@ void test_duplicate_stream() {
 }  // namespace AlgoRandomImpl
 
 TEST(TEST_CATEGORY, Random_XorShift64) {
+#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
+  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
+#endif
```
Comment (masterleinad) [inline algorithms/unit_tests/TestRandom.hpp:547]:
```suggestion
// FIXME KOKKOS_COMPILER_CRAY_LLVM
#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
#endif
```

File: algorithms/unit_tests/TestRandom.hpp [hunk 6]
```diff
@@ -562,6 +566,9 @@ TEST(TEST_CATEGORY, Random_XorShift64) {
 
 TEST(TEST_CATEGORY, Random_XorShift1024_0) {
   using ExecutionSpace = TEST_EXECSPACE;
+#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
+  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
+#endif
```
Comment (masterleinad) [inline algorithms/unit_tests/TestRandom.hpp:571]:
```suggestion
// FIXME KOKKOS_COMPILER_CRAY_LLVM
#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
#endif
```

File: algorithms/unit_tests/TestSort.hpp [hunk 7]
```diff
@@ -209,6 +209,9 @@ void test_sort_integer_overflow() {
 }  // namespace SortImpl
 
 TEST(TEST_CATEGORY, SortUnsignedValueType) {
+#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
+  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
+#endif
```
Comment (masterleinad) [inline algorithms/unit_tests/TestSort.hpp:214]:
```suggestion
// FIXME KOKKOS_COMPILER_CRAY_LLVM
#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
#endif
```

File: algorithms/unit_tests/TestSort.hpp [hunk 8]
```diff
@@ -224,6 +227,9 @@ TEST(TEST_CATEGORY, SortUnsignedValueType) {
 }
 
 TEST(TEST_CATEGORY, SortEmptyView) {
+#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
+  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
+#endif
```
Comment (masterleinad) [inline algorithms/unit_tests/TestSort.hpp:232]:
```suggestion
// FIXME KOKKOS_COMPILER_CRAY_LLVM
#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
#endif
```

File: core/unit_test/TestAtomicOperations_complexfloat.hpp [hunk 9]
```diff
@@ -20,6 +20,9 @@ using namespace TestAtomicOperations;
 
 namespace Test {
 TEST(TEST_CATEGORY, atomic_operations_complexfloat) {
+#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
+  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
+#endif
```
Comment (masterleinad) [inline core/unit_test/TestAtomicOperations_complexfloat.hpp:25]:
```suggestion
// FIXME KOKKOS_COMPILER_CRAY_LLVM
#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
#endif
```

File: core/unit_test/TestAtomicOperations_double.hpp [hunk 10]
```diff
@@ -18,6 +18,9 @@
 
 namespace Test {
 TEST(TEST_CATEGORY, atomic_operations_double) {
+#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
+  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
+#endif
```
Comment (masterleinad) [inline core/unit_test/TestAtomicOperations_double.hpp:23]:
```suggestion
// FIXME KOKKOS_COMPILER_CRAY_LLVM
#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
#endif
```

File: core/unit_test/TestAtomicOperations_float.hpp [hunk 11]
```diff
@@ -18,6 +18,9 @@
 
 namespace Test {
 TEST(TEST_CATEGORY, atomic_operations_float) {
+#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
+  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
+#endif
```
Comment (masterleinad) [inline core/unit_test/TestAtomicOperations_float.hpp:23]:
```suggestion
// FIXME KOKKOS_COMPILER_CRAY_LLVM
#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
#endif
```

File: core/unit_test/TestAtomicOperations_int.hpp [hunk 12]
```diff
@@ -18,6 +18,9 @@
 
 namespace Test {
 TEST(TEST_CATEGORY, atomic_operations_int) {
+#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
+  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
+#endif
```
Comment (masterleinad) [inline core/unit_test/TestAtomicOperations_int.hpp:23]:
```suggestion
// FIXME KOKKOS_COMPILER_CRAY_LLVM
#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
#endif
```

File: core/unit_test/TestAtomicOperations_longint.hpp [hunk 13]
```diff
@@ -18,6 +18,9 @@
 
 namespace Test {
 TEST(TEST_CATEGORY, atomic_operations_long) {
+#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
+  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
+#endif
```
Comment (masterleinad) [inline core/unit_test/TestAtomicOperations_longint.hpp:23]:
```suggestion
// FIXME KOKKOS_COMPILER_CRAY_LLVM
#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
#endif
```

File: core/unit_test/TestAtomicOperations_longlongint.hpp [hunk 14]
```diff
@@ -18,6 +18,9 @@
 
 namespace Test {
 TEST(TEST_CATEGORY, atomic_operations_longlong) {
+#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
+  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
+#endif
```
Comment (masterleinad) [inline core/unit_test/TestAtomicOperations_longlongint.hpp:23]:
```suggestion
// FIXME KOKKOS_COMPILER_CRAY_LLVM
#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
#endif
```

File: core/unit_test/TestAtomicOperations_shared.hpp [hunk 15]
```diff
@@ -19,7 +19,8 @@
 namespace Test {
 
 // FIXME_SYCL This doesn't work yet for SYCL+CUDA
-#if !defined(KOKKOS_ENABLE_SYCL) || defined(KOKKOS_ARCH_INTEL_GPU)
+#if !defined(KOKKOS_ENABLE_OPENMPTARGET) || !defined(KOKKOS_ENABLE_SYCL) || \
+    defined(KOKKOS_ARCH_INTEL_GPU)
```
Comment (masterleinad) [inline core/unit_test/TestAtomicOperations_shared.hpp:23]: Do you mean to disable the test independent of the compiler for `OpenMPTarget`?

File: core/unit_test/TestAtomicOperations_unsignedint.hpp [hunk 16]
```diff
@@ -18,6 +18,9 @@
 
 namespace Test {
 TEST(TEST_CATEGORY, atomic_operations_unsigned) {
+#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
+  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
+#endif
```
Comment (masterleinad) [inline core/unit_test/TestAtomicOperations_unsignedint.hpp:23]:
```suggestion
// FIXME KOKKOS_COMPILER_CRAY_LLVM
#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
#endif
```

File: core/unit_test/TestAtomicOperations_unsignedlongint.hpp [hunk 17]
```diff
@@ -18,6 +18,9 @@
 
 namespace Test {
 TEST(TEST_CATEGORY, atomic_operations_unsignedlong) {
+#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
+  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
+#endif
```
Comment (masterleinad) [inline core/unit_test/TestAtomicOperations_unsignedlongint.hpp:23]:
```suggestion
// FIXME KOKKOS_COMPILER_CRAY_LLVM
#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
#endif
```

File: core/unit_test/TestAtomicOperations_unsignedlonglongint.hpp [hunk 18]
```diff
@@ -18,6 +18,9 @@
 
 namespace Test {
 TEST(TEST_CATEGORY, atomic_operations_unsignedlonglong) {
+#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
+  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
+#endif
```
Comment (masterleinad) [inline core/unit_test/TestAtomicOperations_unsignedlonglongint.hpp:23]:
```suggestion
// FIXME KOKKOS_COMPILER_CRAY_LLVM
#if defined(KOKKOS_COMPILER_CRAY_LLVM) && defined(KOKKOS_ENABLE_OPENMPTARGET)
  GTEST_SKIP() << "known to fail with OpenMPTarget+Cray LLVM";
#endif
```

File: core/unit_test/TestAtomicOperations_shared.hpp (hunk 15 above)
Comment (masterleinad) [reply core/unit_test/TestAtomicOperations_shared.hpp:23]:
```suggestion
// FIXME OPENMPTARGET
#if !defined(KOKKOS_ENABLE_OPENMPTARGET) || !defined(KOKKOS_ENABLE_SYCL) || \
    defined(KOKKOS_ARCH_INTEL_GPU)
```

**Review by masterleinad — COMMENTED** [review]
Goodness! Do you have any idea why these tests are failing only for CrayClang?

File: algorithms/unit_tests/TestBinSortA.hpp (hunk 1 above)
Comment (dalg24) [reply algorithms/unit_tests/TestBinSortA.hpp:222]: The suggested FIXME marker seems inconsistent with our current practice

**Comment by rgayatri23** [comment]
> Goodness! Do you have any idea why these tests are failing only for CrayClang?

Not yet. The `TestAtomic*` tests seem to be all causing ICE for the same issue.
FYI - These tests fail at compile time not runtime so its not a correctness issue but a compiler bug.

File: algorithms/unit_tests/TestBinSortA.hpp (hunk 1 above)
Comment (rgayatri23) [reply algorithms/unit_tests/TestBinSortA.hpp:222]: The FIXME marker should start with `FIXME_OPENMPTARGET - ...` . I will make the changes.

File: core/unit_test/TestAtomicOperations_shared.hpp (hunk 15 above)
Comment (rgayatri23) [reply core/unit_test/TestAtomicOperations_shared.hpp]: No you are right. I fixed the condition on when to skip this test.

File: algorithms/unit_tests/TestBinSortA.hpp [hunk 19]
```diff
@@ -219,6 +219,10 @@ void test_sort_integer_overflow() {
 }  // namespace BinSortSetA
 
 TEST(TEST_CATEGORY, BinSortGenericTests) {
+  // FIXME_OPENMPTARGET - causes ICE with CrayClang compiler
```
Comment (dalg24) [inline algorithms/unit_tests/TestBinSortA.hpp:222]: These comment look wrong to me.  How are these skipping the test help if that was an internal compiler error?
Is that right?  The error was at compile time, not runtime?!

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_Common.hpp [hunk 20]
```diff
@@ -205,7 +205,7 @@ struct ParallelReduceSpecialize<FunctorType, Kokkos::RangePolicy<PolicyArgs...>,
     // architecture in the future.
```
Comment (dalg24) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_Common.hpp:205]: Do you still need this comment?
Also with these changes it looks like you don't need `MAX_ACTIVE_THREADS` any more in which case you should get rid of it.

**Review by crtrott — CHANGES_REQUESTED** [review]
The comment added is definitely wrong. This is not an ICE (ICE stands for Internal Compiler Error, which happens during compile time). I guess you are crashing at runtime?

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_Common.hpp (hunk 20 above)
Comment (rgayatri23) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel_Common.hpp:205]: Yeah got rid of `MAX_ACTIVE_THREADS`. Not needed anymore.

**Comment by rgayatri23** [comment]
Fixed the comments.

**Review by crtrott — APPROVED** [review]

**Comment by dalg24** [comment]
Retest this please

**Comment by dalg24** [comment]
The OpenMPTarget build passed

**Pull request merged**